instrumentation/google.golang.org/grpc/otelgrpc/filters/interceptorinfo.go

```diff
@@ -0,0 +1,102 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package filters // import "go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc/filters"
+
+import (
+	"context"
+
+	"google.golang.org/grpc"
+)
+
+// interceptorType is the flag to define which gRPC interceptor
+// the InterceptorInfo object is.
+type interceptorType uint8
+
+const (
+	unaryClient interceptorType = iota
+	streamClient
+	unaryServer
+	streamServer
+)
+
+// InterceptorInfo is the union of all arguments to four types of
+// gRPC interceptors, except for function types and function arguments
+// of invoker and streamer:
+// * invoker  grpc.UnaryInvoker (UnaryClient)
+// * streamer grpc.Streamer (StreamClient)
+// * stream   grpc.ServerStream (StreamServer)
+// * handler  grpc.UnaryHandler | grpc.StreamHandler (UnaryServer, StreamServer)
+// * req, reply, srv interface{} (UnaryClient, UnaryServer, StreamClient)
+// * cc       *grpc.ClientConn.
+type InterceptorInfo struct {
```

Comment:
It feels a bit strange to make this and NewXInterceptorInfo functions public, as we don't expect users to ever need to use or call them.  Would it be better to leave filters in the otelgrpc package?  It also eliminates the `filters.Filter` stutter for naming.

Reply:
Yes, I totally agree! This is because I created `filters` package to follow the package structure in `otelhttp`. (It has `otelhttp.filters`) I moved around the NewXXXInterceptors and InterceptorInfo back and forth among `otelgrpc`, `otelgrpc/internal` and `otelgrpc/filters` to eliminate unnecessary public members, but I gave up and settled down to the current state. If it's totally ok to move everything under `otelgrpc` and `otelgrpc/internal` directories, then I'm happy to do it.

Reply:
I put the filter implementations in a separate package because most often they'll be referred to by users and `filters.MethodPrefix()` makes more sense to me than `otelhttp.MethodPrefix()`, doubly so for the meta-filters.  `otelhttp.Any()` is less indicative of behavior than `filters.Any()`.  

As for exporting the interceptor info struct, I think it might be worthwhile to do so as without it users cannot implement their own filters.  That would also necessitate exporting its fields.  I think I'm fine with not exporting it for now, but if there's a way to ensure this isn't a one-way door and we can change our minds later that would be great.

Reply:
I think:

* `filters.MethodPrefix` feels better than `otelgrpc.MethodPrefix`
* `filters.Any` feels better than `otelgrpc.Any` (or `otelgrpc.FilterAny`)
* `otelgrpc.Filter` feels better than `filters.Filter`
* `otelgrpc.InterceptorInfo` feels better than `filters.InterceptorInfo`

To get that, we would need to make InterceptorInfo (and its fields) public, but could keep NewXInterceptorInfo private, which I'm OK with.  Would that work @Aneurysm9?